jscomp/syntax/src/res_core.ml

```diff
@@ -982,6 +982,75 @@ let parseRegion p ~grammar ~f =
   Parser.eatBreadcrumb p;
   nodes
 
+let isJsxPropWellFormed p =
+  (* jsx-prop ::= prop = value *)
+  (* Possible tokens after `=` *)
+  (* value ::= *)
+  (*  | True *)
+  (*  | False *)
+  (*  | #Variant *)
+  (*  | `string` *)
+  (*  | [array] *)
+  (*  | list{} *)
+  (*  | () *)
+  (*  | ?value *)
+  (*  | <element /> *)
+  (*  | %raw *)
+  (*  | module(P) *)
+  (*  | string-literal *)
+  (*  | int-literal *)
+  (*  | float-literal *)
+  (*  | variable *)
+  (*  | A.B *)
+  let isPossibleAfterEqual token =
+    match token with
+    | Token.True | False | Hash | Backtick | Lbracket | List | Lparen | Question
+    | LessThan | Percent | Module | String _ | Int _ | Float _ | Lident _
+    | Uident _ ->
+      true
+    | _ -> false
+  in
+  (* jsx-prop ::= prop = {value} *)
+  (* Possible tokens after `}` *)
+  (* prop={{expr}} *)
+  (* prop={value} > ...children </> *)
+  (* prop={value} /> *)
+  (* prop={value} prop2={value2} /> *)
+  (* prop={value} ?prop2 /> *)
+  let isPossibleAfterRbrace token =
+    match token with
+    | Token.Rbrace | GreaterThan | Forwardslash | Lident _ | Question | Eof ->
+      true
+    | _ -> false
+  in
+  let res =
+    Parser.lookahead p (fun state ->
+        match state.Parser.token with
+        (* arrived at k1= *)
+        | Equal -> (
+          Parser.next state;
+          match state.Parser.token with
+          (* arrived at k1={ *)
+          | Lbrace -> (
+            Parser.next state;
+            match state.Parser.token with
+            | Rbrace -> false
+            | _ ->
+              goToClosing Rbrace state;
+              isPossibleAfterRbrace state.Parser.token)
+          (* arrived at k1=v1 *)
+          | token when isPossibleAfterEqual token -> (
+            Parser.next state;
+            match state.Parser.token with
+            (* arrived at k1=v1 =v2 *)
+            | Equal -> false
+            | _ -> true)
+          (* arrived at k1=x *)
+          | _ -> false)
+        | _ -> false)
```

Comment:
When looking ahead, there are two possibilities when we encounter a prop,

1. prop where the value is wrapped with `{}` like `prop={...}`.
   - any expression can live inside this `{}`
3. prop where the value is specified without `{}` like `prop=true`.
    - this can be number literal, string literal, list, array, module, etc

Reply:
I think for the first case, we could just check for `{` (`Lbrace`) but for completeness’s sake, it's probably a good idea to be this explicit.

Reply:
To keep things simple, I would do nothing in the `{` case, and avoid long look aheads searching for `}`.
Then keep the changes and additional code to a minimum.

Reply:
One more case that is not covered is prop punning. If we have just `y` instead of `y=e` and the user types `x=`, then we get `x=y` with no parse error at all. It's also totally not obvious that the intention is to autocomplete.
In other words, this approach does not work at all with prop pinning.

@zth perhaps one can try with a different approach directly in the vscode extension: in case of `x=y` if the cursor is after `=` and `y` is a valid prop name for the component, then infer the intent to autocomplete `x=`. This would cover punned as well as non punned cases. Thoughts?